src/quota_aggregator_impl.cc

```diff
@@ -0,0 +1,248 @@
+
+  *response = lookup.value()->quota_response();
+
+  return ::google::protobuf::util::Status::OK;
+}
+
+// Caches a response from a remote Service Controller AllocateQuota call.
+::google::protobuf::util::Status QuotaAggregatorImpl::CacheResponse(
+    const ::google::api::servicecontrol::v1::AllocateQuotaRequest& request,
+    const ::google::api::servicecontrol::v1::AllocateQuotaResponse& response) {
+  if (!cache_) {
+    return ::google::protobuf::util::Status::OK;
+  }
+
+  string request_signature = GenerateAllocateQuotaRequestSignature(request);
+
+  CacheElem* cache_elem = new CacheElem(response, SimpleCycleTimer::Now());
+  cache_elem->set_signature(request_signature);
+
+  AllocateQuotaCacheRemovedItemsHandler::StackBuffer stack_buffer(this);
+  MutexLock lock(cache_mutex_);
+  AllocateQuotaCacheRemovedItemsHandler::StackBuffer::Swapper swapper(
+      this, &stack_buffer);
+
+  QuotaCache::ScopedLookup lookup(cache_.get(), request_signature);
+  if (lookup.Found()) {
+    if (lookup.value()->is_aggregated()) {
+      // Cached response will be aggregated to the new element for the response
+      cache_elem->Aggregate(lookup.value()->ReturnAllocateQuotaRequestAndClear(
+          service_name_, service_config_id_));
+    }
+  }
+
+  // insert new (aggregated) response to the cache, old one will be removed
+  // internally
+  cache_->Insert(request_signature, cache_elem, 1);
+
+  return ::google::protobuf::util::Status::OK;
+}
+
+// When the next Flush() should be called.
+// Returns in ms from now, or -1 for never
+int QuotaAggregatorImpl::GetNextFlushInterval() {
+  if (!cache_) return -1;
+  return options_.refresh_interval_ms;
+}
+
+// Invalidates expired allocate quota responses.
+// Called at time specified by GetNextFlushInterval().
+::google::protobuf::util::Status QuotaAggregatorImpl::Flush() {
+  AllocateQuotaCacheRemovedItemsHandler::StackBuffer stack_buffer(this);
+  MutexLock lock(cache_mutex_);
+  AllocateQuotaCacheRemovedItemsHandler::StackBuffer::Swapper swapper(
+      this, &stack_buffer);
+
+  if (cache_) {
+    cache_->RemoveExpiredEntries();
+  }
+
+  return Status::OK;
+}
+
+// Flushes out all cached check responses; clears all cache items.
+// Usually called at destructor.
+::google::protobuf::util::Status QuotaAggregatorImpl::FlushAll() {
+  AllocateQuotaCacheRemovedItemsHandler::StackBuffer stack_buffer(this);
+  MutexLock lock(cache_mutex_);
+  AllocateQuotaCacheRemovedItemsHandler::StackBuffer::Swapper swapper(
+      this, &stack_buffer);
+
+  GOOGLE_LOG(INFO) << "Remove all entries of check aggregator.";
+  if (cache_) {
+    cache_->RemoveAll();
+  }
+
+  return Status::OK;
+}
+
+// OnCacheEntryDelete will be called behind the cache_mutex_
+// no need to consider locking at this point
+//
+// if the element is refreshing and aggregated while waiting for the
+// response, tokens should be aggregated
+void QuotaAggregatorImpl::OnCacheEntryDelete(CacheElem* elem) {
+  if (elem->is_aggregated() == true) {
```

Comment:
in what condidtion that "is_aggregated()" flag on? 

When I read this code, I could not follow.

The logic should be:
* for a negative item,  send request with minimum cost, insert a temp positive response
* for a positive item, if not aggregated cost, just delete it. other wise send the aggregated cost to server